tests/check_pip

```diff
@@ -0,0 +1,193 @@
+#
+# This script attempts to enforce spell-checking if the codespell tool is available.
+# It accomplishes the following tasks:
+# 1. ???
+#
+# Usage Summary:
+# ???
+#
+# Exit Code Summary:
+# The script uses the EXIT_CODE variable to track exit conditions:
+# -   0: Successful execution.
+# -   1: General failure.
+# -   2: Coverage combine or XML generation failed.
+# -   3: git ls-tree command failed.
+# -   4: pip-audit failed.
+# -   5: pip-license failed.
+# -  40: Missing valid repository or source structure.
+# - 126: Script already in progress or command not executable.
+# - 129: Received SIGHUP signal.
+# - 130: Received SIGINT signal (Ctrl+C).
+# - 131: Received SIGQUIT signal.
+# - 137: Received SIGABRT signal.
+# - 143: Received SIGTERM signal.
+#
+# The primary goal is to allow linter like spell-checking.
+
+ulimit -t 600
+# setting the path may break brain-dead CI that uses crazy paths
+# PATH="/bin:/sbin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin"
+umask 137
+
+# force utf-8 for spelling
+export LC_CTYPE="${LC_CTYPE:-en_US.UTF-8}"
+
+LOCK_FILE="${TMPDIR:-/tmp}/org.pak.multicast.pip-check-shell"
+EXIT_CODE=1
+
+test -x $(command -v grep) || exit 126 ;
+test -x $(command -v python3) || exit 126 ;
+test -x $(command -v git) || exit 126 ;
+hash -p ./.github/tool_shlock_helper.sh shlock || exit 255 ;
+test -x "$(command -v shlock)" || exit 126 ;
+test -x $(command -v pip-audit) || exit 126 ;
+test -x $(command -v pip-licenses) || exit 126 ;
+
+# Set pip-audit options
+AUDIT_OPTIONS="--progress-spinner off --desc on --requirement"
+
+# Set pip-licenses options
+LICENSE_OPTIONS="--from=mixed"
+
+function cleanup() {
+	rm -f ${LOCK_FILE} 2>/dev/null || : ; wait ;
+	# unset when done
+	unset LOCK_FILE 2>/dev/null || : ;
+	hash -d shlock 2>/dev/null || : ;
+}
+
+if [[ ( $(shlock -f ${LOCK_FILE} -p $$ ) -eq 0 ) ]] ; then
+		EXIT_CODE=0
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 129 ;' SIGHUP || EXIT_CODE=129
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 143 ;' SIGTERM || EXIT_CODE=143
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 131 ;' SIGQUIT || EXIT_CODE=131
+		# SC2173 - https://github.com/koalaman/shellcheck/wiki/SC2173
+		#trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 1 ;' SIGSTOP || EXIT_CODE=7
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 130 ;' SIGINT || EXIT_CODE=130
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true || true ; wait ; exit 137 ;' SIGABRT || EXIT_CODE=137
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit ${EXIT_CODE} ;' EXIT || EXIT_CODE=1
+else
+		# shellcheck disable=SC2046
+		printf "\t%s\n" "Check Setup Scripts Tests Coverage already in progress by "$(head "${LOCK_FILE}") >&2 ;
+		exit 126 ;
+fi
```

Comment:
This is just a 🔨 hack, but a well tested hack.

## TL;DR:
If it were certain all environments would have a `shlock` tool based on the netBSD/Darwin `shlock` implementation I'd agree with your suggestion. However, it is not certain on even common linux flavors like Ubuntu (as used in CI) so at some point I got tired of needing to install and setup a vendor `shlock` tool, especially when it was different from environment to environment, and just wrote a portable shell-script to handle `shlock` type lock-files. please read [.github/tool_shlock_helper.sh](https://github.com/reactive-firewall/multicast/blob/master/.github/tool_shlock_helper.sh) which is hashed to 'shlock' earlier in [this script around line 102](https://github.com/reactive-firewall/multicast/pull/123/files/21f71a7a90d52d27c71fddd30c8094003497e3c2#diff-43d4c8bc7a066c93c6351ff557a683cf3aa67d1c3efd52e83883dcb5e54c7d9fR102-R103). Eventually I'll probably break this tool out into its own separate code repository, but of-course that is, and would be, well outside the scope of this project. 🤷🏻 Over the years I've basically just copied the template version of my 'tool_shlock_helper.sh' tool into projects where I need to ensure a file-locking tool is present.
The caveat of my tool may not behave the same in all cases as an actual `shlock` implementation, the only guarantee is that exit-code 0 will be returned if the lock-file (or by lack of the lock-file) indicates an unlocked state for the given lock-file (there is a global instance lock used as a default lock-file) this use-case also works with a built-in `shlock` but not always the reverse. To avoid the dangers of that reverse case where shlock behaves differently than my 'tool_shlock_helper.sh' the hackish check here only accepts the exit-code if it is indeed equal to 0.
Further as it is used as a locking based critical section solution, the sub-shell `$( ... )` syntax can result in a unique PID value for the actual check process (but also a useful value because it is still a child process and not just random) so this indeed needs to be cleaned up as you observed already. The very next part of the script (assuming the return code of 0 which indicates a successful file-lock now exists) is to `trap` various signals including interrupt, to ensure cleanup of the lock-file (which may be unique to the now expired check sub-shell) however now the script's own shell now effectively inherits hold of the file-lock by this "trapping" (pun intended). From there until the final cleanup (either by a `trap` or by [line 185](https://github.com/reactive-firewall/multicast/blob/feature-150-99-implement-5/tests/check_pip#L185)) the script is in a mutex-like guarded critical section.

In conclusion, while it may not use a `C`-type `flock()` it strikes a balance between universally portable, consistency, and concurrency-safe (as in reliability) which should justify the slight loss of clarity. Please analyzes my response, as a human, I may be mistaken. I look forward to your feedback.

Reply:
Please generate possible documentation for this discussion as if it were a formal, albeit brief, RFC proposal, and render it in two formats: plain-text and GH markdown.

Reply:
This is a great start. Let's continue exploring possible documentation, and improve it. First, consider generalizing the use-case in `check_pip` to shell-scripts, project-wide (of-course not including the custom `tool_shlock_helper.sh` itself as that would be circular). Let's use the codename 'CEP-5' (as in convention enhancement proposal number 5) for the documentation so not to confuse this RFC (request for comment) with the well-known, published RFCs.